File: common/src/main/scala/firelib/utils/HeapQuantile.scala
package firelib.utils

import scala.collection.mutable


class HeapQuantile(val quantile: Double, val length: Int) {

    class Node(val value: Double, var isLeft: Boolean) {}

    private val ring = new mutable.Queue[Node]()

    private val ordering: Ordering[Node] = Ordering.fromLessThan[Node](_.value < _.value)

    private val left = new mutable.TreeSet[Node]()(ordering)
    private val right = new mutable.TreeSet[Node]()(ordering)

    private var cnt = 0

    def count = cnt

    def value: Double = {
        if (left.isEmpty && right.isEmpty) {
            return 0
        }
        if (left.isEmpty) {
            return right.firstKey.value
        }
        if (right.isEmpty) {
            return left.lastKey.value
        }
        return (left.lastKey.value + right.firstKey.value) / 2
    }

    def addMetric(m: Double) = {
        if (ring.size >= length) {
            val node = ring.dequeue()
            if (node.isLeft) {
                left.remove(node)
            }
            else {
                right.remove(node)
            }
        }
        cnt += 1

        if (m > value) {
            var n = new Node(m, false)
            ring += n
            right += n
        }
        else {
            var n = new Node(m, true)
            ring += n
            left += n
        }
        balance()

        if (false) {
            if (left.size > 2) {
                assert(left.firstKey.value <= left.lastKey.value)
            }
            if (right.size > 2) {
                assert(right.firstKey.value <= right.lastKey.value)
            }
            if (left.size > 2 && right.size > 2) {
                assert(left.lastKey.value <= right.firstKey.value)
            }
        }
    }

    private def balance() {
        val diff = left.size / quantile - right.size / (1 - quantile)
        if (diff > 0) {
            val amDiff = (left.size - 1) / quantile - (right.size + 1) / (1 - quantile)
            if (math.abs(amDiff) < math.abs(diff)) {
                var leftMax = left.lastKey
                left.remove(leftMax)
                right += leftMax
                leftMax.isLeft = false

            }
        }
        else {
            val amDiff = (left.size + 1) / quantile - (right.size - 1) / (1 - quantile)
            if (math.abs(amDiff) < math.abs(diff)) {
                var rightMin = right.firstKey
                right.remove(rightMin)
                left += rightMin
                rightMin.isLeft = true
            }
        }
    }
}

File: common/src/main/scala/firelib/utils/HeapQuantile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using C5;


namespace QuantLib.Rolling.HeapQuantile
{

    public class HeapQuantile
    {
        private readonly double quantile;

        private readonly TreeBag<Node> left = new TreeBag<Node>(new NodeComparer());
        private readonly TreeBag<Node> right = new TreeBag<Node>(new NodeComparer());
        private readonly CircularQueue<Node> ring;

        private int cnt = 0;
        readonly int length;
        public HeapQuantile(double quantile, int count)
        {
            this.quantile = quantile;
            ring = new CircularQueue<Node>(count);
            length = count;
        }

        class Node
        {
            public readonly double Value;
            public bool Left;

            public Node(double value, bool left)
            {
                this.Value = value;
                this.Left = left;
            }
        }

        class NodeComparer : IComparer<Node>
        {
            public int Compare(Node n0, Node n1)
            {
                return n0.Value.CompareTo(n1.Value);
            }

        }

        public int Count
        {
            get { return cnt; }
        }

        public double Quant
        {
            get { return quantile; }
        }

        public double GetValue()
        {
            if (left.IsEmpty && right.IsEmpty)
            {
                return 0;
            }
            if (left.IsEmpty)
            {
                return right.FindMin().Value;
            }
            if (right.IsEmpty)
            {
                return left.FindMax().Value;
            }

            return (left.FindMax().Value + right.FindMin().Value) / 2;
        }

        public void AddMetric(double m)
        {
            if (ring.Count >= length)
            {
                Node node = ring.Dequeue();
                if (node.Left)
                {
                    left.Remove(node);
                }
                else
                {
                    right.Remove(node);
                }
            }
            cnt++;

            var val = GetValue();

            if (m > val)
            {
                var n = new Node(m, false);
                ring.Enqueue(n);
                right.Add(n);
            }
            else
            {
                var n = new Node(m, true);
                ring.Enqueue(n);
                left.Add(n);
            }
            Balance();
        }

        void Balance()
        {
            var diff = left.Count / quantile - right.Count / (1 - quantile);
            if (diff > 0)
            {
                var amDiff = (left.Count - 1) / quantile - (right.Count + 1) / (1 - quantile);
                if (Math.Abs(amDiff) < diff)
                {
                    var leftMax = left.DeleteMax();
                    right.Add(leftMax);
                    leftMax.Left = false;
                }
            }
            else
            {
                var amDiff = (left.Count + 1) / quantile - (right.Count - 1) / (1 - quantile);
                if (Math.Abs(amDiff) < -diff)
                {
                    var rightMin = right.DeleteMin();
                    left.Add(rightMin);
                    rightMin.Left = true;
                }
            }
        }
    }
}